bitcoin/src/blockdata/witness.rs

```diff
@@ -147,19 +147,15 @@ crate::internal_macros::define_extension_trait! {
         ///
         /// See [`Script::is_p2tr`] to check whether this is actually a Taproot witness.
         fn tapscript(&self) -> Option<&Script> {
-            self.last().and_then(|last| {
-                // From BIP341:
-                // If there are at least two witness elements, and the first byte of
-                // the last element is 0x50, this last element is called annex a
-                // and is removed from the witness stack.
-                if self.len() >= 3 && last.first() == Some(&TAPROOT_ANNEX_PREFIX) {
-                    self.nth(self.len() - 3).map(Script::from_bytes)
-                } else if self.len() >= 2 {
-                    self.nth(self.len() - 2).map(Script::from_bytes)
-                } else {
-                    None
-                }
-            })
+            if self.is_empty() {
+                return None;
+            }
+
+            if self.taproot_annex().is_some() {
+                self.third_to_last().map(Script::from_bytes)
```

Comment:
Take it or leave it, but I think it's better to unwrap here and below:

```suggestion
                Some(Script::from_bytes(self.third_to_last().unwrap()))
```

Since we expect `third_to_last()` and `second_to_lat()` to always return `Some` in this case. This makes that explicit, and if for some reason they return `None`, due to a bug, we panic instead of returning incorrect results.

Reply:
IIUC we don't guarantee that this is script spend, the docs say this just returns the third to last element if it exists. So we cannot unwrap here (unless we want to explicitly panic if this is called with a keypath spend, which is an unrelated change IMO).

Reply:
Whoops! I wasn't paying attention to the fact that we hadn't checked if were were in a script path.